packages/labs/task/src/task.ts

```diff
@@ -400,16 +423,93 @@ export class Task<
         throw new Error(`Unexpected status: ${this.status}`);
     }
   }
-
-  private _argsDirty(args?: T) {
-    const prev = this._previousArgs;
-    this._previousArgs = args;
-    return Array.isArray(args) && Array.isArray(prev)
-      ? args.length === prev.length && args.some((v, i) => notEqual(v, prev[i]))
-      : args !== prev;
-  }
 }
 
 type MaybeReturnType<F> = F extends (...args: unknown[]) => infer R
   ? R
   : undefined;
+
+export const shallowArrayEquals = <T extends ReadonlyArray<unknown>>(
+  oldArgs: T,
+  newArgs: T
+) =>
+  oldArgs === newArgs ||
+  (oldArgs.length === newArgs.length &&
+    oldArgs.every((v, i) => !notEqual(v, newArgs[i])));
+
+export const deepArrayEquals = <T extends ReadonlyArray<unknown>>(
+  oldArgs: T,
+  newArgs: T
+) =>
+  oldArgs === newArgs ||
+  (oldArgs.length === newArgs.length &&
+    oldArgs.every((v, i) => deepEquals(v, newArgs[i])));
+
+const objectValueOf = Object.prototype.valueOf;
+const objectToString = Object.prototype.toString;
+const {keys: objectKeys} = Object;
+const {isArray} = Array;
+
+export const deepEquals = (a: unknown, b: unknown): boolean => {
+  if (Object.is(a, b)) {
+    return true;
+  }
+
+  if (
+    a !== null &&
+    b !== null &&
+    typeof a === 'object' &&
+    typeof b === 'object'
+  ) {
+    // Object must have the same prototype / constructor
+    if (a.constructor !== b.constructor) {
+      return false;
+    }
+
+    // Arrays must have the same length and recursively equal items
+    if (isArray(a)) {
+      if (a.length !== (b as Array<unknown>).length) {
+        return false;
+      }
+      return a.every((v, i) => deepEquals(v, (b as Array<unknown>)[i]));
+    }
+
+    // Defer to custom valueOf implementations. This handles Dates which return
+    // ms since epoch: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/valueOf
+    if (a.valueOf !== objectValueOf) {
+      return a.valueOf() === b.valueOf();
+    }
+
+    // Defer to custom toString implementations. This should handle
+    // TrustedTypes, URLs, and such. This might be a bit risky, but
+    // fast-deep-equals does it.
+    if (a.toString !== objectToString) {
+      return a.toString() === b.toString();
```

Comment:
What about Map, Set?

Reply:
I can add those. I don't think it's actually too many bytes.